File: src/main/org/nlogo/api/ShapeList.java
package org.nlogo.api;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public strictfp class ShapeList {

  public static final String DEFAULT_SHAPE_NAME = "default";

  public static boolean isDefaultShapeName(String name) {
    return name.equals(DEFAULT_SHAPE_NAME);
  }

  ///

  private final Map<String, Shape> shapes = new HashMap<String, Shape>();

  public ShapeList(Shape... shapes) {
    for (Shape shape : shapes) {
      add(shape);
    }
  }

  public Shape shape(String name) {
    Shape shape = shapes.get(name);
    return shape == null
        ? shapes.get(DEFAULT_SHAPE_NAME)
        : shape;
  }

  // Returns vector of the list of shapes available to the current model
  public List<Shape> getShapes() {
    List<Shape> currentShapes = new ArrayList<Shape>();
    for (Shape nextShape : shapes.values()) {
      // don't add the default shape here, we will add it later so that it
      // is at the top of the list
      if (!isDefaultShapeName(nextShape.getName())) {
        currentShapes.add(nextShape);
      }
    }
    List<Shape> sortedShapes = sortShapes(currentShapes);
    // make sure that the shape with the name DEFAULT_SHAPE_NAME is at the
    // top of the list.
    sortedShapes.add(0, shapes.get(DEFAULT_SHAPE_NAME));
    return sortedShapes;
  }

  public static final List<Shape> sortShapes(List<Shape> unsortedShapes) {
    List<Shape> result = new ArrayList<Shape>(unsortedShapes);
    Collections.sort
        (result,
            new Comparator<Shape>() {
              // compare two instances of shapemodel so we can return a
              // sorted shape list
              public int compare(Shape shape1, Shape shape2) {
                return shape1.getName().compareTo(shape2.getName());
              }
            });
    return result;
  }

  // Returns a set of the names of all available shapes
  public Set<String> getNames() {
    return shapes.keySet();
  }

  // Returns true when a shape with the given name is already available to the current model
  public boolean exists(String name) {
    return shapes.containsKey(name);
  }

  // Clears the list of shapes currently available
  public void replaceShapes(Collection<Shape> newShapes) {
    shapes.clear();
    addAll(newShapes);
  }

  // Adds a new shape to the ones currently available for use
  public Shape add(Shape newShape) {
    return shapes.put(newShape.getName(), newShape);
  }

  // Adds a collection of shapes to the ones currently available for use
  public void addAll(Collection<Shape> collection) {
    for (Shape shape : collection) {
      add(shape);
    }
  }

  // Removes a shape from those currently in use
  public Shape removeShape(Shape shapeToRemove) {
    return shapes.remove(shapeToRemove.getName());
  }

}

File: headless/src/main/org/nlogo/api/ShapeList.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.api

import java.util.{ Collection => JCollection, List => JList, Set => JSet }
import collection.JavaConverters._

object ShapeList {
  val DefaultShapeName = "default"
  def isDefaultShapeName(name: String) =
    name == DefaultShapeName
  def sortShapes(unsortedShapes: JList[Shape]): JList[Shape] =
    collection.mutable.ArrayBuffer(unsortedShapes.asScala: _*)
      .sortBy(_.getName)
      .asJava
}

class ShapeList(val kind: AgentKind, _shapes: Seq[Shape]) {

  def this(kind: AgentKind) = this(kind, Seq())

  private val shapes = collection.mutable.HashMap[String, Shape]()

  _shapes.foreach(add)

  import ShapeList._

  def shape(name: String): Shape =
    shapes.get(name).getOrElse(shapes(DefaultShapeName))

  /** Returns vector of the list of shapes available to the current model */
  def getShapes: JList[Shape] = {
    // leave out the default shape for now; we will add it later so that it is at the top of the list
    val currentShapes =
      shapes.values.toSeq.filterNot(s => isDefaultShapeName(s.getName))
    val sortedShapes = new java.util.ArrayList[Shape]
    sortedShapes.addAll(sortShapes(currentShapes.asJava))
    // make sure that the shape with the name DefaultShapeName is at the top of the list.
    sortedShapes.add(0, shapes(DefaultShapeName))
    sortedShapes
  }

  /** Returns a set of the names of all available shapes */
  def getNames: JSet[String] =
    shapes.keySet.asJava

  /** Returns true when a shape with the given name is already available to the current model */
  def exists(name: String) =
    shapes.contains(name)

  /** Clears the list of shapes currently available */
  def replaceShapes(newShapes: JCollection[Shape]) {
    shapes.clear()
    addAll(newShapes)
  }

  /** Adds a new shape to the ones currently available for use */
  def add(newShape: Shape): Shape = {
    val replaced = shapes.get(newShape.getName).orNull
    shapes(newShape.getName) = newShape
    replaced
  }

  /** Adds a collection of shapes to the ones currently available for use */
  def addAll(collection: JCollection[Shape]) {
    collection.asScala.foreach(add)
  }

  /** Removes a shape from those currently in use */
  def removeShape(shapeToRemove: Shape) = {
    val removed = shapes.get(shapeToRemove.getName).orNull
    shapes -= shapeToRemove.getName
    removed
  }

}
